### PR TITLE
improve usb errata 5 feature's documentation.

### PR DESCRIPTION
**Pull request by ithinuel**

### CODE REVIEW

**Comment by dlkj** [comment]
LGTM, make this much clearer.

**Comment by ithinuel** [comment]
@9names @jannic I added a bypass for the workaround if it is running on rev B2 and above.
@dlkj would you mind testing it with your system?

**Comment by jannic** [comment]
> @9names @jannic I added a bypass for the workaround if it is running on rev B2 and above.

I'm not 100% sure if I like this. This can lead to situations were a firmware works perfectly for the developer, who happens to use a B2 stepping, and breaks for users with older chips.

In that sense, I'd prefer something like https://github.com/rp-rs/rp-hal/pull/608: Make it a guaranteed panic if you have the e5 workaround enabled and forget to initialize bank0 before USB.
(That code is completely untested - please just treat it as an idea.)

**Comment by jannic** [comment]
And I forgot to say: The update of the doc comment is great!

**Comment by ithinuel** [comment]
> I'm not 100% sure if I like this. This can lead to situations were a firmware works perfectly for the developer, who happens to use a B2 stepping, and breaks for users with older chips.
> 
> In that sense, I'd prefer something like https://github.com/rp-rs/rp-hal/pull/608: Make it a guaranteed panic if you have the e5 workaround enabled and forget to initialize bank0 before USB.
> (That code is completely untested - please just treat it as an idea.)

Oh I didn't think about this indeed, you're totally right.
Do you mind if I incorporate your suggestions to this PR?

I think bypassing the workaround is a good thing on B2 revs to avoid messing needlessly with the hardware but checking correct implementation with the requirements of the feature make sense too.

**Comment by jannic** [comment]
> Do you mind if I incorporate your suggestions to this PR?

Of course not, go ahead!

**Review by jannic — APPROVED** [review]
I added some (minor) comments, but overall, this looks good.

File: rp2040-hal/src/usb/errata5.rs [hunk 1]
```diff
@@ -57,6 +53,18 @@ impl Errata5State {
             }
         }
     }
+
+    /// Make sure bank0 is out of reset, which is necessary for the rp2040-e5 workaround.
+    /// If it is not, panic.
+    pub fn check_bank0_reset() {
```
[the remark this answers is not in the document]
Comment (ithinuel) [reply rp2040-hal/src/usb/errata5.rs:59]: The `errata5` module is private so thing made "pub" in this module are not accessible publicly.

File: rp2040-hal/src/usb/errata5.rs [hunk 2]
```diff
@@ -29,15 +29,11 @@ impl Errata5State {
         let pac = crate::pac::Peripherals::steal();
         match self {
             Self::WaitEndOfReset => {
-                if pac.USBCTRL_REGS.sie_status.read().line_state().is_se0() {
+                if pac.SYSINFO.chip_id.read().revision().bits() >= 2 {
```
[the remark this answers is not in the document]
Comment (ithinuel) [reply rp2040-hal/src/usb/errata5.rs:32]: For consistency I think it should indeed.

File: rp2040-hal/src/usb/errata5.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jannic) [reply rp2040-hal/src/usb/errata5.rs:59]: Good point, I didn't notice that.

**Review by jannic — APPROVED** [review]

**Comment by dlkj** [comment]
I've not had the time to check this yet. I'll find some time this week probably.

**Comment by dlkj** [comment]
I can't reproduce the panic now using the same pico and a very similar code base. LGTM.

**Pull request merged**